### PR TITLE
Add tutorial for android java

### LINKED ISSUE #82: 📚 Documentation: Android tutorial using Java
### 💭 Description

While the Appwrite team will write an App using Android Kotlin, we do not have a tutorial on Java. For community members that are fluent in Android development using Java, we'd love to see contributions here.

### Goal
Interactive tutorial for beginners to build their first functioning app using Android (Java) and Appwrite.

### Content
1. The end product should be a functional app
2. Step by step from setup. If there will be test data, provide a Server SDK script with instructions.
3. Format should be similar to ` /docs/tutorials/sveltekit`
4. Should contain the following content
    1. install dependencies
    2. make first request 
    3. implement auth
    4. intro to databases
    5. queries and paginations
    6. storage (optional)
    7. realtime (optional)
    8. permissions (optional)

### How to contribute
Pitch an outline detailing each page you plan to add and headings within, and ask to be assigned. Open a PR **after** the outline has been approved.

The Appwrite team will collaborate with you on the outline and PR.

Thanks for taking the time to make our documentation better 🙏

### 👀 Have you spent some time to check if this issue has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/appwrite/.github/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by letsintegreat**
## What does this PR do?

Closes #82 

This PR adds a tutorial to `/docs/tutorials/android`

The tutorial gives a step-by-step guide to creating a recipe-sharing app - Recipewrite with Android and Java.

## Test Plan

The final code can be found [here](https://github.com/letsintegreat/Recipewrite), if you would like to test it.

### Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?

Yes.

### CODE REVIEW

**Comment by gewenyu99** [comment]
Hi, can you give use the originals (unedited) of all your screenshots in light and dark more? Uncropped 16:9.

For console screenshots 1400x900

We'll also have to redo your other images, but we'll accept this without visuals.

File: src/routes/docs/tutorials/android/step-5/+page.markdoc [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+In **Attributes** tab, create two new attributes.
+
+{% only_dark %}
+![Create project screen](/images/docs/tutorials/dark/recipewrite-collection-attributes.png)
+{% /only_dark %}
+{% only_light %}
+![Create project screen](/images/docs/tutorials/recipewrite-collection-attributes.png)
+{% /only_light %}
+
+- **name**
```
Comment (gewenyu99) [inline src/routes/docs/tutorials/android/step-5/+page.markdoc:26]: Please use a table.

File: src/routes/docs/tutorials/android/step-3/+page.markdoc [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+{% only_dark %}
+![Add a platform](/images/docs/quick-starts/dark/add-platform.png)
+{% /only_dark %}
+{% only_light %}
+![Add a platform](/images/docs/quick-starts/add-platform.png)
+{% /only_light %}
+
+You can skip the optional steps.
+
+# Connect your app with appwrite {% #connect-your-app-with-appwrite %}
```
Comment (gewenyu99) [inline src/routes/docs/tutorials/android/step-3/+page.markdoc:35]: Please capitalize Appwrite

File: src/routes/docs/tutorials/android/step-3/+page.markdoc [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+Create **Appwrite.java**, this will handle our app's interaction with Appwrite server.
```
Comment (gewenyu99) [inline src/routes/docs/tutorials/android/step-3/+page.markdoc:37]: In what path? Always mention a path

File: src/routes/docs/tutorials/android/step-3/+page.markdoc [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+```java
+public class Appwrite {
```
Comment (gewenyu99) [inline src/routes/docs/tutorials/android/step-3/+page.markdoc:40]: This is missing `package com.example.recipewrite;` right?

File: src/routes/docs/tutorials/android/step-4/+page.markdoc [hunk 5]
```diff
@@ -0,0 +1,509 @@
+---
+layout: tutorial
+title: Authentication
+description: Learn to build an Android app with no backend code using an Appwrite backend.
+step: 4
+difficulty: beginner
+---
+
+# User Flow {% #user-flow %}
```
Comment (gewenyu99) [inline src/routes/docs/tutorials/android/step-4/+page.markdoc:9]: Anyway to simplify this page? This is way too much, maybe break this up, but I think this is borderline overkill.

File: src/routes/docs/tutorials/android/step-2/+page.markdoc [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+![New Project](/images/docs/tutorials/recipewrite-new-project.png)
+
+# Add dependencies {% #add-dependencies %}
+
+1. Open **app/build.gradle** file
+
+2. Find `dependencies` block, and add the following dependencies along with existing ones
+    ```
+    implementation 'io.appwrite:sdk-for-android:4.0.0'
```
Comment (gewenyu99) [inline src/routes/docs/tutorials/android/step-2/+page.markdoc:29]: This should be double quotes " with ()

File: src/routes/docs/tutorials/android/step-3/+page.markdoc [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+![Add a platform](/images/docs/quick-starts/add-platform.png)
+{% /only_light %}
+
+You can skip the optional steps.
+
+# Connect your app with appwrite {% #connect-your-app-with-appwrite %}
+
+Create **Appwrite.java**, this will handle our app's interaction with Appwrite server.
+
+```java
```
Comment (gewenyu99) [inline src/routes/docs/tutorials/android/step-3/+page.markdoc:39]: Also missing imports

File: src/routes/docs/tutorials/android/step-4/+page.markdoc [hunk 8]
```diff
@@ -0,0 +1,509 @@
...
+
+This is the user flow that we are going to achieve in this page
+
+![user-flow](/images/docs/tutorials/recipewrite-user-flow.png)
+
+Keeping the userflow in mind, let us start by initializing three new empty activities -
+
+- **LoginActivity**
```
Comment (gewenyu99) [inline src/routes/docs/tutorials/android/step-4/+page.markdoc:17]: This list is unnecessary.

File: src/routes/docs/tutorials/android/step-4/+page.markdoc [hunk 9]
```diff
@@ -0,0 +1,509 @@
...
+layout: tutorial
+title: Authentication
+description: Learn to build an Android app with no backend code using an Appwrite backend.
+step: 4
+difficulty: beginner
+---
+
+# User Flow {% #user-flow %}
+
+This is the user flow that we are going to achieve in this page
```
Comment (gewenyu99) [inline src/routes/docs/tutorials/android/step-4/+page.markdoc:11]: Missing punctuation.

File: src/routes/docs/tutorials/android/step-4/+page.markdoc [hunk 10]
```diff
@@ -0,0 +1,509 @@
...
+
+![user-flow](/images/docs/tutorials/recipewrite-user-flow.png)
+
+Keeping the userflow in mind, let us start by initializing three new empty activities -
```
Comment (gewenyu99) [inline src/routes/docs/tutorials/android/step-4/+page.markdoc:15]: Please use standard English punctuation here.

File: src/routes/docs/tutorials/android/step-4/+page.markdoc [hunk 11]
```diff
@@ -0,0 +1,509 @@
...
+
+- **LoginActivity**
+- **RegisterActivity**
+- **HomeActivity**
+
+# Appwrite methods {% #appwrite-methods %}
+
+Let us implement some utility methods in **Appwrite.java**
+
+- **onGetAccount**
```
Comment (gewenyu99) [inline src/routes/docs/tutorials/android/step-4/+page.markdoc:25]: This should not be a list.

File: src/routes/docs/tutorials/android/step-4/+page.markdoc [hunk 12]
```diff
@@ -0,0 +1,509 @@
...
+            context.startActivity(intent);
+            ((Activity) context).finish();
+        }));
+    } catch (Exception e) {
+        e.printStackTrace();
+    }
+}
+```
+
+- **onLogin**
```
Comment (gewenyu99) [inline src/routes/docs/tutorials/android/step-4/+page.markdoc:56]: Again should not be a list.

File: src/routes/docs/tutorials/android/step-4/+page.markdoc [hunk 13]
```diff
@@ -0,0 +1,509 @@
...
+# Appwrite methods {% #appwrite-methods %}
+
+Let us implement some utility methods in **Appwrite.java**
+
+- **onGetAccount**
+
+It will be used in **MainActivity** to check if the user is already logged in or not.
+
+```java
+public static void onGetAccount(Context context) {
```
Comment (gewenyu99) [inline src/routes/docs/tutorials/android/step-4/+page.markdoc:30]: Missing imports. These should not be in Appwrite.js tbh. Why is there UI logic in Appwrite.java, which is just a service provider for Appwrite client?

File: src/routes/docs/tutorials/android/step-4/+page.markdoc [hunk 14]
```diff
@@ -0,0 +1,509 @@
...
+        case R.id.register_button:
+            /* Do something */
+            break;
+    }
+}
+```
+
+We can simply use our Appwrite method to login the user, if **login_button** is pressed -
+
+```java
```
Comment (gewenyu99) [inline src/routes/docs/tutorials/android/step-4/+page.markdoc:291]: The way you present this code is really hard to follow, I can't tell where this code goes, what imports I need where, what the file structure of the project is.

Can we simplify this so that we can have a more minimalistic authentication component in 1-2 smaller files?

**Review by gewenyu99 — CHANGES_REQUESTED** [review]
Hey, I took my first look. Thank you so much for contributing to Appwrite, appreciate the effort you put in!

I have a few preliminary comments. Please take a look. 

Love the idea, concerned about how complicated the auth components look :P

**Comment by gewenyu99** [comment]
Hey,

Due to time constraints, I'm going to mark this PR `hacktoberfest-accepted` for now so you get DO's Hacktoberfest rewards. We'll continue to work with you on this issue for review and merge.

When it is merged, we'll contact you for Appwrite-specific Hacktoberfest swag.

Thanks for helping us improve Appwrite!

File: src/routes/docs/tutorials/android/step-5/+page.markdoc (hunk 1 above)
Comment (letsintegreat) [reply src/routes/docs/tutorials/android/step-5/+page.markdoc]: Done.

File: src/routes/docs/tutorials/android/step-5/+page.markdoc [hunk 15]
```diff
@@ -0,0 +1,328 @@
...
+In **Indexes** tab, create a new index. This will be important for search feature of our app.
+
+{% only_dark %}
+![Create project screen](/images/docs/tutorials/dark/recipewrite-collection-indexes.png)
+{% /only_dark %}
+{% only_light %}
+![Create project screen](/images/docs/tutorials/recipewrite-collection-indexes.png)
+{% /only_light %}
+
+- **recipe_index**
```
[the remark this answers is not in the document]
Comment (letsintegreat) [reply src/routes/docs/tutorials/android/step-5/+page.markdoc]: Done.

File: src/routes/docs/tutorials/android/step-2/+page.markdoc [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+description: Learn to build an Android app with no backend code using an Appwrite backend.
+step: 2
+difficulty: beginner
+---
+
+# Initialize Android Studio project {% #initialize-android-studio-project %}
+
+1. Fire up your Android Studio
+
+2. Go to **File / New / New Project**
```
[the remark this answers is not in the document]
Comment (letsintegreat) [reply src/routes/docs/tutorials/android/step-2/+page.markdoc]: Done, please check if I need to change it further.

File: src/routes/docs/tutorials/android/step-2/+page.markdoc [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+layout: tutorial
+title: Initialize project
+description: Learn to build an Android app with no backend code using an Appwrite backend.
+step: 2
+difficulty: beginner
+---
+
+# Initialize Android Studio project {% #initialize-android-studio-project %}
+
+1. Fire up your Android Studio
```
[the remark this answers is not in the document]
Comment (letsintegreat) [reply src/routes/docs/tutorials/android/step-2/+page.markdoc]: Fixed.

File: src/routes/docs/tutorials/android/step-2/+page.markdoc (hunk 6 above)
[the remark this answers is not in the document]
Comment (letsintegreat) [reply src/routes/docs/tutorials/android/step-2/+page.markdoc]: Tried something else, LMK if it is still weird.

File: src/routes/docs/tutorials/android/step-2/+page.markdoc [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+
+![Empty Activity](/images/docs/tutorials/recipewrite-empty-activity.png)
+
+4. Name your project **Recipewrite**, select Java language, and minimum SDK would be **23**
+
+![New Project](/images/docs/tutorials/recipewrite-new-project.png)
+
+# Add dependencies {% #add-dependencies %}
+
+1. Open **app/build.gradle** file
```
[the remark this answers is not in the document]
Comment (letsintegreat) [reply src/routes/docs/tutorials/android/step-2/+page.markdoc]: Fenced file names with ` throughout.

File: src/routes/docs/tutorials/android/step-2/+page.markdoc [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+
+2. Find `dependencies` block, and add the following dependencies along with existing ones
+    ```
+    implementation 'io.appwrite:sdk-for-android:4.0.0'
+    ```
+    
+3. Update `compileSdk` and `targetSdk` to **34** in the same file
```
[the remark this answers is not in the document]
Comment (letsintegreat) [reply src/routes/docs/tutorials/android/step-2/+page.markdoc]: Fixed.

File: src/routes/docs/tutorials/android/step-3/+page.markdoc [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+{% only_dark %}
+![Create project screen](/images/docs/quick-starts/dark/create-project.png)
+{% /only_dark %}
+{% only_light %}
+![Create project screen](/images/docs/quick-starts/create-project.png)
+{% /only_light %}
+
+If this is your first time using Appwrite, create an account and create your first project.
+
+Give your project a new name, and set **Project ID** to `recipewrite`.
```
[the remark this answers is not in the document]
Comment (letsintegreat) [reply src/routes/docs/tutorials/android/step-3/+page.markdoc]: Done.

File: src/routes/docs/tutorials/android/step-2/+page.markdoc [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+difficulty: beginner
+---
+
+# Initialize Android Studio project {% #initialize-android-studio-project %}
+
+1. Fire up your Android Studio
+
+2. Go to **File / New / New Project**
+
+3. Select **Empty Activity**
```
[the remark this answers is not in the document]
Comment (letsintegreat) [reply src/routes/docs/tutorials/android/step-2/+page.markdoc]: Ah, thanks for pointing that out, I have made the necessary changes in the text, but haven't yet replaced the image, as you are anyways going to replace these images, I will provide the new screenshot to you here.

File: src/routes/docs/tutorials/android/step-3/+page.markdoc (hunk 2 above)
Comment (letsintegreat) [reply src/routes/docs/tutorials/android/step-3/+page.markdoc]: Done this change throughout.

File: src/routes/docs/tutorials/android/step-3/+page.markdoc (hunk 3 above)
Comment (letsintegreat) [reply src/routes/docs/tutorials/android/step-3/+page.markdoc]: Fixed.

File: src/routes/docs/tutorials/android/step-3/+page.markdoc (hunk 4 above)
Comment (letsintegreat) [reply src/routes/docs/tutorials/android/step-3/+page.markdoc:38]: I am just providing the entire classes and methods and not caring about package names, imports, etc. That is not an issue because of modern text editors, and I believe they would just add to the noise in our tutorial. LMK if you feel otherwise.

File: src/routes/docs/tutorials/android/step-3/+page.markdoc (hunk 7 above)
Comment (letsintegreat) [reply src/routes/docs/tutorials/android/step-3/+page.markdoc:37]: I am just providing the entire classes and methods and not caring about package names, imports, etc. That is not an issue because of modern text editors, and I believe they would just add to the noise in our tutorial. LMK if you feel otherwise.

File: src/routes/docs/tutorials/android/step-4/+page.markdoc (hunk 5 above)
Comment (letsintegreat) [reply src/routes/docs/tutorials/android/step-4/+page.markdoc:9]: Thanks for pointing this out. I have revamped the entire authentication user-flow, and have scrapped `RegisterActivity`. Now, there's only one activity for authentication, `AuthActivity`. Please check the new user-flow, and LMK what you think.

File: src/routes/docs/tutorials/android/step-2/+page.markdoc (hunk 6 above)
Comment (letsintegreat) [reply src/routes/docs/tutorials/android/step-2/+page.markdoc]: Not sure what you mean here, can you please elaborate? Sorry.

File: src/routes/docs/tutorials/android/step-4/+page.markdoc (hunk 8 above)
Comment (letsintegreat) [reply src/routes/docs/tutorials/android/step-4/+page.markdoc]: Agreed and removed.

File: src/routes/docs/tutorials/android/step-4/+page.markdoc (hunk 9 above)
Comment (letsintegreat) [reply src/routes/docs/tutorials/android/step-4/+page.markdoc]: I have fixed all punctuation errors.

File: src/routes/docs/tutorials/android/step-4/+page.markdoc (hunk 10 above)
Comment (letsintegreat) [reply src/routes/docs/tutorials/android/step-4/+page.markdoc]: Fixed.

File: src/routes/docs/tutorials/android/step-4/+page.markdoc (hunk 11 above)
Comment (letsintegreat) [reply src/routes/docs/tutorials/android/step-4/+page.markdoc]: Agreed and removed.

File: src/routes/docs/tutorials/android/step-4/+page.markdoc (hunk 12 above)
Comment (letsintegreat) [reply src/routes/docs/tutorials/android/step-4/+page.markdoc]: Agreed and removed.

File: src/routes/docs/tutorials/android/step-4/+page.markdoc (hunk 13 above)
Comment (letsintegreat) [reply src/routes/docs/tutorials/android/step-4/+page.markdoc]: You are right, I apologize for having a weak understanding here. I have removed all of the UI logic from `Appwrite.java`. However, I still think that having `Appwrite.java` is helpful as it will serve as the one place where all IDs and endpoints related to Appwrite are kept. LMK if you feel otherwise.

File: src/routes/docs/tutorials/android/step-4/+page.markdoc (hunk 14 above)
Comment (letsintegreat) [reply src/routes/docs/tutorials/android/step-4/+page.markdoc]: I have simplified authentication to be in just one activity now.

**Comment by letsintegreat** [comment]
@gewenyu99 Thanks a lot for such a thorough review, I really appreciate the time and effort you have put into reviewing my PR. I have addressed all the requested changes and would like you to review this PR again. Thanks!

**Comment by gewenyu99** [comment]
I'm gonna merge this into a branch and see what I can do to make this up to our docs' standards. Thank you!

**Comment by letsintegreat** [comment]
@gewenyu99 my discord username is letsintegreat

**Comment by gewenyu99** [comment]
Will reach out ASAP. Preparing lots of forms for everyone! Appreciate your patience!
